platform/ke02/export/aux/uart.hpp

```diff
@@ -0,0 +1,348 @@
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::rx_done))
+               == (ctx::inited | ctx::rx_done));
+
+    bus_ctx.rx = rx;
+    bus_ctx.rx_sz = size;
+    bus_ctx.rx_idx = 0;
+}
+
+template<uart_channel ch>
+void uart<ch>::set_tx(const uint8_t *tx, size_t size)
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done))
+               == (ctx::inited | ctx::tx_done));
+
+    bus_ctx.tx.buf = tx;
+    bus_ctx.tx_sz = size;
+    bus_ctx.tx_idx = 0;
+
+    bus_ctx.status &= ~ctx::fill;
+}
+
+template<uart_channel ch>
+void uart<ch>::set_tx(size_t size, uint8_t fill_byte)
+{
+    // TODO: implement
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done))
+               == (ctx::inited | ctx::tx_done));
+
+    bus_ctx.tx.fill_byte = fill_byte;
+    bus_ctx.tx_sz = size;
+
+    bus_ctx.status |= ctx::fill;
+}
+
+template<uart_channel ch>
+void uart<ch>::set_handler(const bus_handler &handler)
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done | ctx::rx_done))
+               == (ctx::inited | ctx::tx_done | ctx::rx_done));
+
+    bus_ctx.h = handler;
+}
+
+template<uart_channel ch>
+void uart<ch>::reset_buffers()
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done | ctx::rx_done))
+               == (ctx::inited | ctx::tx_done | ctx::rx_done));
+
+    // If buffers are not set then fill mode must be disabled to avoid
+    // ambiguity.
+    bus_ctx.status &= ~ctx::fill;
+    bus_ctx.tx.buf = nullptr;
+    bus_ctx.rx = nullptr;
+}
+
+template<uart_channel ch>
+void uart<ch>::reset_handler()
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done | ctx::rx_done))
+               == (ctx::inited | ctx::tx_done | ctx::rx_done));
+
+    bus_ctx.h = stub_handler;
+}
+
+template<uart_channel ch>
+err uart<ch>::do_xfer()
+{
+    auto &bus_ctx = get_ctx();
+
+    ecl_assert((bus_ctx.status & (ctx::inited | ctx::tx_done | ctx::rx_done))
+               == (ctx::inited | ctx::tx_done | ctx::rx_done));
+
+    // At least one of direction should be enabled
+    ecl_assert(bus_ctx.tx.buf || bus_ctx.rx);
+
+    if ((bus_ctx.status & ctx::fill) || bus_ctx.tx.buf) {
+        UART_Type * ut = pick_sysctl();
+
+        if (bus_ctx.status & ctx::fill) {
+            for(size_t i = 0; i < bus_ctx.tx_sz; i++) {
+                UART_PutChar(ut, bus_ctx.tx.fill_byte);
+            }
+        } else {
+            while(bus_ctx.tx_idx < bus_ctx.tx_sz) {
+                UART_PutChar(ut, bus_ctx.tx.buf[bus_ctx.tx_idx]);
+                bus_ctx.tx_idx++;
+            }   
+        }
+
+        bus_ctx.status |= ctx::tx_done;
+    }
+
+    if (bus_ctx.rx) {
+        // TODO: Implement RX mechanism.
+        bus_ctx.status |= ctx::rx_done;
+    }
+
+    bus_ctx.h(bus_channel::meta, bus_event::tc, 0);
```

Comment:
FYI: to able to correctly process RX stream and do not miss any byte from remote device, we need to implement IRQ-based data retrieval (this is a subject to further tasks)